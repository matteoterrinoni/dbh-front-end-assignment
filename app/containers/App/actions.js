/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_USERS,
  LOAD_USERS_SUCCESS,
  LOAD_USERS_ERROR,
  LOAD_USER,
  LOAD_USER_SUCCESS,
  LOAD_USER_ERROR,
  EDIT_USER,
  EDIT_USER_ERROR,
  EDIT_USER_SUCCESS,
} from './constants';

/**
 * Load the users, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_USERS
 */
export function loadUsers() {
  return {
    type: LOAD_USERS,
  };
}

/**
 * Dispatched when the users are loaded by the request saga
 *
 * @param  {array} users The users data
 *
 * @return {object}      An action object with a type of LOAD_USERS_SUCCESS passing the users
 */
export function usersLoaded(users) {
  return {
    type: LOAD_USERS_SUCCESS,
    users,
  };
}

/**
 * Dispatched when loading the users fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_USERS_ERROR passing the error
 */
export function usersLoadingError(error) {
  return {
    type: LOAD_USERS_ERROR,
    error,
  };
}

/**
 * Load the users, this action starts the request saga
 * @param  {string} uid The error
 *
 * @return {object} An action object with a type of LOAD_USERS
 */
export function loadUser(uid) {
  return {
    type: LOAD_USER,
    uid,
  };
}

/**
 * Dispatched when the users are loaded by the request saga
 *
 * @param  {object} user The user data
 *
 * @return {object}      An action object with a type of LOAD_USERS_SUCCESS passing the users
 */
export function userLoaded(user) {
  return {
    type: LOAD_USER_SUCCESS,
    user,
  };
}

/**
 * Dispatched when loading the users fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_USERS_ERROR passing the error
 */
export function userLoadingError(error) {
  return {
    type: LOAD_USER_ERROR,
    error,
  };
}

/**
 * Load the users, this action starts the request saga
 *
 * @param  {object} user The error
 *
 * @return {object} An action object with a type of LOAD_USERS
 */
export function editUser(user) {
  return {
    type: EDIT_USER,
    user,
  };
}

/**
 * Dispatched when the users are loaded by the request saga
 *
 * @param  {object} user The user data
 *
 * @return {object}      An action object with a type of LOAD_USERS_SUCCESS passing the user
 */
export function userEdited(user) {
  return {
    type: EDIT_USER_SUCCESS,
    user,
  };
}

/**
 * Dispatched when loading the users fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_USERS_ERROR passing the error
 */
export function userEditingError(error) {
  return {
    type: EDIT_USER_ERROR,
    error,
  };
}
