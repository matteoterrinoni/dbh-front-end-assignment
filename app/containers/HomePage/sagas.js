/**
 * Gets the users of the user from Github
 */

import { take, call, put, cancel, takeLatest } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import { LOAD_USERS } from 'containers/App/constants';
import { usersLoaded, usersLoadingError } from 'containers/App/actions';

import request from 'utils/request';

import Api from 'containers/App/model/api';

/**
 * users request/response handler
 */
export function* getUsers() {
  // Select username from store
  const requestURL = `${Api.basePath}users`;

  try {
    // Call our request helper (see 'utils/request')
    const users = yield call(request, requestURL);
    yield put(usersLoaded(users));
  } catch (err) {
    const e = yield err;
    yield put(usersLoadingError(e));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* usersData() {
  // Watches for LOAD_USERS actions and calls getRepos when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  const watcher = yield takeLatest(LOAD_USERS, getUsers);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

// Bootstrap sagas
export default [usersData];
