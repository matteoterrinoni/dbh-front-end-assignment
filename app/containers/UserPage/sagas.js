import { take, call, put, cancel, takeLatest } from 'redux-saga/effects';
import { LOCATION_CHANGE } from 'react-router-redux';
import {
  userLoaded,
  userLoadingError,
  userEdited,
  userEditingError,
} from 'containers/App/actions';
import { EDIT_USER, LOAD_USER } from 'containers/App/constants';
import request from 'utils/request';

/**
 * Github user request/response handler
 */
export function* getUser({ uid }) {
  // Select username from store
  const requestURL = `http://localhost:5000/users/${uid}`;

  try {
    // Call our request helper (see 'utils/request')
    const user = yield call(request, requestURL);
    yield put(userLoaded(user));
  } catch (err) {
    const e = yield err;
    yield put(userLoadingError(e));
  }
}

export function* editUser({ user }) {
  // Select username from store
  const requestURL = `http://localhost:5000/users/${user.id}`;

  let edited;

  try {
    // Call our request helper (see 'utils/request')
    edited = yield call(request, requestURL, {
      method: 'PUT',
      body: JSON.stringify(user),
      headers: {
        'Content-Type': 'Application/json',
      },
    });
    yield put(userEdited(edited));
  } catch (err) {
    const e = yield err;
    yield put(userEditingError(e));
  }
}

/**
 * Root saga manages watcher lifecycle
 */
export function* userData() {
  // Watches for LOAD_USER actions and calls getUser when one comes in.
  // By using `takeLatest` only the result of the latest API call is applied.
  // It returns task descriptor (just like fork) so we can continue execution
  const watcher = yield takeLatest(LOAD_USER, getUser);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

export function* editUserData() {
  // Watches for EDIT_USER and calls editUser
  const watcher = yield takeLatest(EDIT_USER, editUser);

  // Suspend execution until location changes
  yield take(LOCATION_CHANGE);
  yield cancel(watcher);
}

// Bootstrap sagas
export default [userData, editUserData];
