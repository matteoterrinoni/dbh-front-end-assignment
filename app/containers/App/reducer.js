/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */

import { combineReducers } from 'redux-immutable';
import { List } from 'immutable';

import {
  LOAD_USERS_SUCCESS,
  LOAD_USERS,
  LOAD_USERS_ERROR,
  LOAD_USER_SUCCESS,
  LOAD_USER,
  LOAD_USER_ERROR,
  EDIT_USER,
  EDIT_USER_SUCCESS,
  EDIT_USER_ERROR,
} from './constants';

import { Users } from '../UserPage/model';

function loading(state = false, action) {
  switch (action.type) {
    case LOAD_USERS:
    case LOAD_USER:
    case EDIT_USER:
      return true;
    case LOAD_USERS_SUCCESS:
    case LOAD_USER_SUCCESS:
    case EDIT_USER_SUCCESS:
      return false;
    case LOAD_USERS_ERROR:
    case LOAD_USER_ERROR:
    case EDIT_USER_ERROR:
      return false;
    default:
      return state;
  }
}

function error(state = false, action) {
  switch (action.type) {
    case LOAD_USERS:
    case LOAD_USER:
    case EDIT_USER:
      return false;
    case LOAD_USERS_SUCCESS:
    case LOAD_USER_SUCCESS:
    case EDIT_USER_SUCCESS:
      return false;
    case LOAD_USERS_ERROR:
    case LOAD_USER_ERROR:
    case EDIT_USER_ERROR:
      if (action.error.constructor === Object) {
        return Object.keys(action.error).length === 0 ? false : action.error;
      }
      return action.error || false;
    default:
      return state;
  }
}

function users(state = false, action) {
  if (List.isList(state)) {
    // normalize the state from the server
    state = state.toJS(); // eslint-disable-line no-param-reassign
  }

  const usersState = state || [];

  switch (action.type) {
    case LOAD_USERS:
      return false;
    case LOAD_USERS_SUCCESS:
      return action.users;
    case LOAD_USER_SUCCESS:
      return Users(usersState).set(action.user).users;
    case EDIT_USER_SUCCESS:
      return Users(usersState).set(action.user).users;
    default:
      return state;
  }
}

const data = combineReducers({
  users,
});

export default combineReducers({
  loading,
  error,
  data,
});
