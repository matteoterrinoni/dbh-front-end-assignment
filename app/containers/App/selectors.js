/**
 * The global state selectors
 */

import { createSelector } from 'reselect';
import { Users } from '../UserPage/model';

const selectGlobal = (state) => state.get('global');

const makeSelectLoading = () =>
  createSelector(selectGlobal, (globalState) => globalState.get('loading'));

const makeSelectError = () =>
  createSelector(selectGlobal, (globalState) => globalState.get('error'));

const makeSelectUsers = () =>
  createSelector(selectGlobal, (globalState) =>
    globalState.getIn(['data', 'users'])
  );

const makeSelectUser = (uid) =>
  createSelector(selectGlobal, (globalState) =>
    Users(globalState.getIn(['data', 'users'])).find(uid)
  );

const makeSelectLocationState = () => {
  let prevRoutingState;
  let prevRoutingStateJS;

  return (state) => {
    const routingState = state.get('route'); // or state.route

    if (!routingState.equals(prevRoutingState)) {
      prevRoutingState = routingState;
      prevRoutingStateJS = routingState.toJS();
    }

    return prevRoutingStateJS;
  };
};

export {
  selectGlobal,
  makeSelectLoading,
  makeSelectError,
  makeSelectUsers,
  makeSelectUser,
  makeSelectLocationState,
};
