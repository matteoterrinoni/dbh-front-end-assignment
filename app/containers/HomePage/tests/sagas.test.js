/**
 * Tests for HomePage sagas
 */

import { cancel, take, put, takeLatest } from 'redux-saga/effects';
import { createMockTask } from 'redux-saga/lib/utils';

import { LOCATION_CHANGE } from 'react-router-redux';

import { LOAD_USERS } from 'containers/App/constants';
import { usersLoaded, usersLoadingError } from 'containers/App/actions';

import { getUsers, usersData } from '../sagas';

/* eslint-disable redux-saga/yield-effects */
describe('getUsers Saga', () => {
  let getUsersGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getUsersGenerator = getUsers();

    const selectDescriptor = getUsersGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot([]);
  });

  it('should dispatch the usersLoaded action if it requests the data successfully', () => {
    const response = [
      {
        id: 1,
      },
      {
        id: 2,
      },
    ];

    const putDescriptor = getUsersGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(usersLoaded(response)));
  });

  it('should call the usersLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const error = getUsersGenerator.throw(response).value;
    const putDescriptor = getUsersGenerator.next(error).value;
    expect(putDescriptor).toEqual(put(usersLoadingError(response)));
  });
});

describe('usersDataSaga Saga', () => {
  const usersDataSaga = usersData();
  const mockedTask = createMockTask();

  it('should start task to watch for LOAD_USERS action', () => {
    const takeLatestDescriptor = usersDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_USERS, getUsers));
  });

  it('should yield until LOCATION_CHANGE action', () => {
    const takeDescriptor = usersDataSaga.next(mockedTask).value;
    expect(takeDescriptor).toEqual(take(LOCATION_CHANGE));
  });

  it('should cancel the forked task when LOCATION_CHANGE happens', () => {
    const cancelDescriptor = usersDataSaga.next().value;
    expect(cancelDescriptor).toEqual(cancel(mockedTask));
  });
});
