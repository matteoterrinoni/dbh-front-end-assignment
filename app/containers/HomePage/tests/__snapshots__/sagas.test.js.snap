// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:5000/users",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
